tools/src/gencerts/Main.hs

```diff
@@ -29,92 +25,64 @@ data Command = Command
-                (NonEmpty.head $ serverAltNames $ tlsServer tlsConfig, "")
-
-        (caKey, caCert) <-
-            genCertificate caDesc
-
-        case certOutDir caDesc of
-            Nothing  -> return ()
-            Just dir -> writeCredentials (dir </> caName) (caKey, caCert)
-
-        forM_ descs $ \desc@CertDescription{..} -> do
-            (key, cert) <- genCertificate desc
-            validateSHA256 caCert certChecks (serverHost, serverPort) cert >>= failIfReasons
-            writeCredentials (certOutDir </> certFilename) (key, cert)
-            writeCertificate (certOutDir </> caName)       caCert
-
-
--- | Generate & sign a certificate from a certificate description
-genCertificate
-    :: CertDescription IO PublicKey PrivateKey filename
-    -> IO (PrivateKey, SignedCertificate)
-genCertificate CertDescription{..} = do
-    ((pub, priv), now) <- (,) <$> certGenKeys <*> dateCurrent
-
-    let CertConfiguration{..} = certConfiguration
-    let cert = Certificate
-            { certVersion      = 2
-            , certSerial       = fromIntegral certSerial
-            , certSignatureAlg = signAlgRSA256
-            , certValidity     = (addMinutes (-1) now, addDays certExpiryDays now)
-            , certPubKey       = PubKeyRSA pub
-            , certExtensions   = Extensions (Just certExtensions)
-            , certIssuerDN     = certIssuer
-            , certSubjectDN    = certSubject
-            }
-
-    (priv,) <$> signCertificate certSigningKey cert
-  where
-    addDays :: Int -> DateTime -> DateTime
-    addDays n time@(DateTime dtDate _) =
-        time { dtDate = dateAddPeriod dtDate (mempty { periodDays = n }) }
-
-    addMinutes :: Int -> DateTime -> DateTime
-    addMinutes n time =
-        timeAdd time (Minutes $ fromIntegral n)
+main = do
+    cmd <- execParser $
+        info (cmdParser <**> helper)
+        (  fullDesc
+        <> header "X.509 Certificates Generation"
+        <> progDesc "Pure Haskell 'replacement' for OpenSSL to generate certificates for a TLS Private Key Infrastructure"
+        )
+
+    tlsConfig <-
+        decodeConfigFile (configKey cmd) (configFile cmd)
+
+    (caDesc, descs) <-
+        fromConfiguration tlsConfig (outDirectories cmd) genRSA256KeyPair <$> genRSA256KeyPair
+
+    let caName =
+            certFilename caDesc
+
+    let (serverHost, serverPort) =
+            (NonEmpty.head $ serverAltNames $ tlsServer tlsConfig, "")
+
+    (caKey, caCert) <-
+        genCertificate caDesc
+
+    case certOutDir caDesc of
+        Nothing  -> return ()
+        Just dir -> writeCredentials (dir </> caName) (caKey, caCert)
+
+    forM_ descs $ \desc -> do
+        (key, cert) <- genCertificate desc
+        failIfReasons =<< validateSHA256
+            caCert
+            (certChecks desc)
+            (serverHost, serverPort)
+            cert
+        writeCredentials (certOutDir desc </> certFilename desc) (key, cert)
+        writeCertificate (certOutDir desc </> caName) caCert
```

Comment:
Removing `{-# RecordWildCards #-}` extension and defining `cmdParser` at a higher level for readability. All indentation has therefore shifted.